Sources/Workspace/UserToolchain.swift

```diff
@@ -97,6 +97,81 @@ public final class UserToolchain: Toolchain {
         return runtime
     }
 
+    private static func _hostToolchainSearchPaths() -> [() -> [AbsolutePath]] {
+        /// Searches for the Swift instance seen by the shell.
+        func generalSearch() -> [AbsolutePath] {
+            return (Process.findExecutable("swift")?.parentDirectory.parentDirectory).map({ [$0] }) ?? []
+        }
+        /// Searches for the Swift instance owned by Xcode.
+        func xcodeSearch() -> [AbsolutePath] {
+            let process = Process(args: "xcrun", "--find", "swift")
+            do {
+                try process.launch()
+                let result = try process.waitUntilExit()
+                guard result.exitStatus == .terminated(code: 0) else {
+                    return []
+                }
+                var pathString = try result.utf8Output()
+                if pathString.last == "\n" { pathString.removeLast() }
+                let path = AbsolutePath(pathString).parentDirectory.parentDirectory
+                return [path]
+            } catch {
+                return []
+            }
+        }
+        /// Searches for the Swift instance owned by the Swift Version Manager.
+        /// (https://github.com/kylef/swiftenv#swift-version-manager)
```

Comment:
I don't think it's a good idea to bake in third-party projects into libSwiftPM without discussion with the Swift team. I'd say we leave it out of this PR and discuss `swiftenv` separately on the forums if you really think SwiftPM should use it.

Reply:
To be clear, I don’t think SwiftPM should *use* thirty‐party projects. I do think SwiftPM should *defend against* them.

That means:

1. SwiftPM should make a reasonable attempt to ensure its own integrity. For example, with Xcode installed, and without a toolchain validation heuristic, the package manager library built from source says: “error: toolchain is invalid: could not find the `swiftc` at expected path”. The user can attempt to verify that by running `$ swiftc --version`, but will come to the conclusion that the toolchain is just fine. You and I both know that it is really Xcode’s fault for hacking in its own shims. But to every user out there, it looks like it’s SwiftPM’s fault. `swiftenv` does almost the exact same thing as Xcode, and sees wide use on Linux. It results in the exact same error message. I do think it would be reasonable to emit an error saying “swiftenv has interfered in the swift toolchain; please reinstall swift properly”. That way it doesn’t look like the problem came from SwiftPM and it is clear how to fix it. Ultimately I deemed it easier to silently play along than to correctly diagnose the problem—doing so requires even more knowledge of the workings of the third party tool.

2. SwiftPM should not document anywhere that this little piece of code is here. It is only an implementation detail there for security reasons to help prevent outside influences from pulling the rug out from under it. 

3. SwiftPM should always assume its own intended installation methods first, which are direct download and Xcode right now. This is another point where the search reversal possibility mentioned above runs afoul.

***

You are welcome to share my thinking with the Swift team.

If you decide that it is better to throw an error if interference is detected—which is fine—, then I am forced to accept that responsibility in my client code. That in turn means the client code will always have to handle all the rest of the lookup anyway, because it needs the resolved location before it can pass it into SwiftPM’s API. That in turn means none of the functionality in this pull request will be usable for me, and with my motivation gone, I will not continue investing time in this particular pull request.

It would be helpful to know your decision promptly, since it will affect the direction of the next pull request on my to‐do list. I was going to send in a pull request reducing the effort needed to load manifests and package graphs down to a single line instead of the dozens required now. If `hostDestination()` just works, then it can be done by merely filling in the initializers at each level with default parameters. If `hostDestination()` will require the toolchain path, then every type will need a new initializer that accepts the toolchain path and constructs everything from there. That is a big difference, so I don’t want to start until I hear word from you.

***

(This may interest to you, @kylef.)

Reply:
I understand the problems here but a) we need permission from the project lead when using new OSS in the Swift project b) I think we can solve this in a better way where we expose APIs for clients to integrate external lookup tools like `swiftenv`. That way, we don't have to bake in different kinds of tools in SwiftPM. It does mean that clients will need to do some extra configuration but that seems reasonable to me.

Reply:
Deal. Should I move on a request that fills in as many default parameters as *are* available, or do you want to think about the whole general design for a while first?

Reply:
I've has a suggested approach. I've opened https://github.com/apple/swift-package-manager/pull/2166 to demonstrate what I am thinking. Let's continue conversation there?